<img src="./unal.png" align=left width="150" height="150"></img>

$\;$
---

<h2 align="center">AGREGATION</h2>

A continuación se muestra un conjunto de datos pertenecientes a un restaurante, en este se muestra los detalles de la factura del cliente. Entre las características se destaca:

* **total_bill**: Precio de la factura (puede ser del plato).
* **tip**: Propina.
* **sex**: Sexo del cliente.
* **smoker**: Si fuma o no fuma.
* **day**: Dia de la semana.

In [31]:
import pandas as pd
import numpy as np
import seaborn as sns
sns.__version__

'0.11.1'

In [16]:
pd.options.display.float_format = '{:,.3f}'.format

In [17]:
df=sns.load_dataset('tips')
df

total_bill   tip     sex smoker   day    time  size
0        16.990 1.010  Female     No   Sun  Dinner     2
1        10.340 1.660    Male     No   Sun  Dinner     3
2        21.010 3.500    Male     No   Sun  Dinner     3
3        23.680 3.310    Male     No   Sun  Dinner     2
4        24.590 3.610  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239      29.030 5.920    Male     No   Sat  Dinner     3
240      27.180 2.000  Female    Yes   Sat  Dinner     2
241      22.670 2.000    Male    Yes   Sat  Dinner     2
242      17.820 1.750    Male     No   Sat  Dinner     2
243      18.780 3.000  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [18]:
df.describe(include='all')

total_bill     tip   sex smoker  day    time    size
count      244.000 244.000   244    244  244     244 244.000
unique         NaN     NaN     2      2    4       2     NaN
top            NaN     NaN  Male     No  Sat  Dinner     NaN
freq           NaN     NaN   157    151   87     176     NaN
mean        19.786   2.998   NaN    NaN  NaN     NaN   2.570
std          8.902   1.384   NaN    NaN  NaN     NaN   0.951
min          3.070   1.000   NaN    NaN  NaN     NaN   1.000
25%         13.348   2.000   NaN    NaN  NaN     NaN   2.000
50%         17.795   2.900   NaN    NaN  NaN     NaN   2.000
75%         24.127   3.562   NaN    NaN  NaN     NaN   3.000
max         50.810  10.000   NaN    NaN  NaN     NaN   6.000

Se puede observar que el costo promedio de un plato esta alrededor de 19 dólares, la propina alrededor de 3 USD. En cuanto a las variables categoricas se logra observar que:

* Existen dos variables de sexo.
* Existen dos opciones de respuesta en cuanto a fumar y no fumar.
* Existe registrado 4 dias de la semana..


In [19]:
df['day'].value_counts()

Sat     87
Sun     76
Thur    62
Fri     19
Name: day, dtype: int64

In [20]:
df['day'].value_counts()  /  df['day'].value_counts().sum()   * 100   #observando conteos como porcentajes.

Sat    35.656
Sun    31.148
Thur   25.410
Fri     7.787
Name: day, dtype: float64

In [22]:
df.groupby('sex').mean()

total_bill   tip  size
sex                           
Male        20.744 3.090 2.631
Female      18.057 2.833 2.460

La propina de los hombres es alrededor de 2.6USD y el de las mujeres de 2.4USD. Además los platos comprados por los hombres es de 20USD y el de las mujeres de 18USD.

### Porcentaje de propina

Considerando la especialidad del plato y su precio se determina un porcentaje de propina con respecto al valor dela factura:

In [24]:
df['prct_tip']=df['tip']/df['total_bill']
df

total_bill   tip     sex smoker   day    time  size  prct_tip
0        16.990 1.010  Female     No   Sun  Dinner     2     0.059
1        10.340 1.660    Male     No   Sun  Dinner     3     0.161
2        21.010 3.500    Male     No   Sun  Dinner     3     0.167
3        23.680 3.310    Male     No   Sun  Dinner     2     0.140
4        24.590 3.610  Female     No   Sun  Dinner     4     0.147
..          ...   ...     ...    ...   ...     ...   ...       ...
239      29.030 5.920    Male     No   Sat  Dinner     3     0.204
240      27.180 2.000  Female    Yes   Sat  Dinner     2     0.074
241      22.670 2.000    Male    Yes   Sat  Dinner     2     0.088
242      17.820 1.750    Male     No   Sat  Dinner     2     0.098
243      18.780 3.000  Female     No  Thur  Dinner     2     0.160

[244 rows x 8 columns]

In [25]:
df.groupby('sex').mean()

total_bill   tip  size  prct_tip
sex                                     
Male        20.744 3.090 2.631     0.158
Female      18.057 2.833 2.460     0.166

A qui se puede destacar que los hombres y mujeres aportan un 15% del valor de sus pagos y lo destinan a propinas.

In [26]:
df.groupby('sex').median()

total_bill   tip  size  prct_tip
sex                                     
Male        18.350 3.000     2     0.153
Female      16.400 2.750     2     0.156

In [27]:
df.groupby('sex')[['prct_tip']].describe()

prct_tip                                          
          count  mean   std   min   25%   50%   75%   max
sex                                                      
Male    157.000 0.158 0.065 0.036 0.121 0.153 0.186 0.710
Female   87.000 0.166 0.054 0.056 0.140 0.156 0.194 0.417

In [28]:
df.groupby('sex')[['prct_tip','total_bill']].describe()

prct_tip                                           total_bill         \
          count  mean   std   min   25%   50%   75%   max      count   mean   
sex                                                                           
Male    157.000 0.158 0.065 0.036 0.121 0.153 0.186 0.710    157.000 20.744   
Female   87.000 0.166 0.054 0.056 0.140 0.156 0.194 0.417     87.000 18.057   

                                                
         std   min    25%    50%    75%    max  
sex                                             
Male   9.246 7.250 14.000 18.350 24.710 50.810  
Female 8.009 3.070 12.750 16.400 21.520 44.300

In [32]:
#función paa convertir euros a dolares

def mean_eur2usd(x):
    return np.mean(x)*1.12

In [34]:
mean_eur2usd([100,102])

113.12

In [38]:
df.groupby('sex')[['total_bill','prct_tip']].apply(mean_eur2usd)   #aplicando una función con apply

total_bill  prct_tip
sex                         
Male        23.233     0.177
Female      20.224     0.186

In [40]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].apply(mean_eur2usd)     #clasificando por sexo y por tiempo

total_bill  prct_tip
sex    time                        
Male   Lunch       20.214     0.186
       Dinner      24.037     0.174
Female Lunch       18.300     0.182
       Dinner      21.519     0.190

In [43]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].apply(np.std) 

total_bill  prct_tip
sex    time                        
Male   Lunch        7.832     0.045
       Dinner       9.423     0.069
Female Lunch        7.393     0.034
       Dinner       8.123     0.063

### Crreando función lambda

In [44]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].apply(lambda x:  np.mean(x)*1.12) 

total_bill  prct_tip
sex    time                        
Male   Lunch       20.214     0.186
       Dinner      24.037     0.174
Female Lunch       18.300     0.182
       Dinner      21.519     0.190

#### colocando mas deuna función usar aggregate (también agg)

In [46]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].aggregate([np.mean, np.max])

total_bill        prct_tip      
                    mean   amax     mean  amax
sex    time                                   
Male   Lunch      18.048 41.190    0.166 0.266
       Dinner     21.461 50.810    0.155 0.710
Female Lunch      16.339 43.110    0.162 0.259
       Dinner     19.213 44.300    0.169 0.417

In [47]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].agg([np.mean, np.max])

total_bill        prct_tip      
                    mean   amax     mean  amax
sex    time                                   
Male   Lunch      18.048 41.190    0.166 0.266
       Dinner     21.461 50.810    0.155 0.710
Female Lunch      16.339 43.110    0.162 0.259
       Dinner     19.213 44.300    0.169 0.417

In [50]:
nuev_dict={'tip':[min, max], 'total_bill':[np.mean, mean_eur2usd]}    #este diccionario tiene un conjunto de funciones
nuev_dict

{'tip': [<function min>, <function max>],
 'total_bill': [<function numpy.mean(a, axis=None, dtype=None, out=None, keepdims=<no value>)>,
  <function __main__.mean_eur2usd(x)>]}

In [52]:
df.groupby(['sex','time'])[['total_bill','tip']].agg(nuev_dict)

tip        total_bill             
                min    max       mean mean_eur2usd
sex    time                                       
Male   Lunch  1.440  6.700     18.048       20.214
       Dinner 1.000 10.000     21.461       24.037
Female Lunch  1.250  5.170     16.339       18.300
       Dinner 1.000  6.500     19.213       21.519

### CREANDO FILTRO POR CATEGORIAS CON MEDIA MAYOR A 40

In [53]:
def f_filter(x):
    return mean_eur2usd(x['total_bill'].mean()) > 20

In [56]:
df_filtered = df.groupby(['sex','time']).filter(f_filter)
df_filtered

total_bill   tip     sex smoker   day    time  size  prct_tip
0        16.990 1.010  Female     No   Sun  Dinner     2     0.059
1        10.340 1.660    Male     No   Sun  Dinner     3     0.161
2        21.010 3.500    Male     No   Sun  Dinner     3     0.167
3        23.680 3.310    Male     No   Sun  Dinner     2     0.140
4        24.590 3.610  Female     No   Sun  Dinner     4     0.147
..          ...   ...     ...    ...   ...     ...   ...       ...
239      29.030 5.920    Male     No   Sat  Dinner     3     0.204
240      27.180 2.000  Female    Yes   Sat  Dinner     2     0.074
241      22.670 2.000    Male    Yes   Sat  Dinner     2     0.088
242      17.820 1.750    Male     No   Sat  Dinner     2     0.098
243      18.780 3.000  Female     No  Thur  Dinner     2     0.160

[209 rows x 8 columns]

anteriormente se elimino todas las categorias donde el costo promedio del plato en un determinado grupo sea mayor a 20

In [57]:
df_filtered.groupby(['sex','time']).count()

total_bill  tip  smoker  day  size  prct_tip
sex    time                                                
Male   Lunch           33   33      33   33    33        33
       Dinner         124  124     124  124   124       124
Female Lunch            0    0       0    0     0         0
       Dinner          52   52      52   52    52        52

anteriormente se muestra 0 en las mujeres en hora del almuerzo

---

In [61]:
df['ones']=1
df

total_bill   tip     sex smoker   day    time  size  prct_tip  ones
0        16.990 1.010  Female     No   Sun  Dinner     2     0.059     1
1        10.340 1.660    Male     No   Sun  Dinner     3     0.161     1
2        21.010 3.500    Male     No   Sun  Dinner     3     0.167     1
3        23.680 3.310    Male     No   Sun  Dinner     2     0.140     1
4        24.590 3.610  Female     No   Sun  Dinner     4     0.147     1
..          ...   ...     ...    ...   ...     ...   ...       ...   ...
239      29.030 5.920    Male     No   Sat  Dinner     3     0.204     1
240      27.180 2.000  Female    Yes   Sat  Dinner     2     0.074     1
241      22.670 2.000    Male    Yes   Sat  Dinner     2     0.088     1
242      17.820 1.750    Male     No   Sat  Dinner     2     0.098     1
243      18.780 3.000  Female     No  Thur  Dinner     2     0.160     1

[244 rows x 9 columns]

Conteo de personas que fuman en cada categoria

In [64]:
df_g = df.groupby(['sex','smoker'])['ones'].sum()
df_g 

sex     smoker
Male    Yes       60
        No        97
Female  Yes       33
        No        54
Name: ones, dtype: int64

In [65]:
df_g = df.groupby(['sex','smoker'])[['ones']].sum()
df_g 

ones
sex    smoker      
Male   Yes       60
       No        97
Female Yes       33
       No        54

In [66]:
df_g.groupby(level=0).apply(

lambda x:
    x/ x.sum()*100
)

ones
sex    smoker       
Male   Yes    38.217
       No     61.783
Female Yes    37.931
       No     62.069

---
generando intervalos en variables numéricas

In [67]:
pd.cut(df['total_bill'],bins=3)

0       (3.022, 18.983]
1       (3.022, 18.983]
2      (18.983, 34.897]
3      (18.983, 34.897]
4      (18.983, 34.897]
             ...       
239    (18.983, 34.897]
240    (18.983, 34.897]
241    (18.983, 34.897]
242     (3.022, 18.983]
243     (3.022, 18.983]
Name: total_bill, Length: 244, dtype: category
Categories (3, interval[float64]): [(3.022, 18.983] < (18.983, 34.897] < (34.897, 50.81]]

In [68]:
pd.cut(df['total_bill'],bins=3).unique()

[(3.022, 18.983], (18.983, 34.897], (34.897, 50.81]]
Categories (3, interval[float64]): [(3.022, 18.983] < (18.983, 34.897] < (34.897, 50.81]]

In [69]:
pd.cut(df['total_bill'],bins=3).value_counts()

(3.022, 18.983]     140
(18.983, 34.897]     88
(34.897, 50.81]      16
Name: total_bill, dtype: int64

No obstante, se crean categorias con  respecto a la variable valor_factura donde existen 140 personas en el intervalo (3.022, 18.983], 88 personas en el intervalo de (18.983, 34.897] dolares y 16 e el intervalo (34.897, 50.81].

Definiendo el ancho de los intervalos con:

In [71]:
pd.cut(df['total_bill'],bins=[3,18,35,60]).value_counts()

(3, 18]     127
(18, 35]    101
(35, 60]     16
Name: total_bill, dtype: int64

In [72]:
df['bin_total']=pd.cut(df['total_bill'],bins=[3,18,35,60])
df

total_bill   tip     sex smoker   day    time  size  prct_tip  ones  \
0        16.990 1.010  Female     No   Sun  Dinner     2     0.059     1   
1        10.340 1.660    Male     No   Sun  Dinner     3     0.161     1   
2        21.010 3.500    Male     No   Sun  Dinner     3     0.167     1   
3        23.680 3.310    Male     No   Sun  Dinner     2     0.140     1   
4        24.590 3.610  Female     No   Sun  Dinner     4     0.147     1   
..          ...   ...     ...    ...   ...     ...   ...       ...   ...   
239      29.030 5.920    Male     No   Sat  Dinner     3     0.204     1   
240      27.180 2.000  Female    Yes   Sat  Dinner     2     0.074     1   
241      22.670 2.000    Male    Yes   Sat  Dinner     2     0.088     1   
242      17.820 1.750    Male     No   Sat  Dinner     2     0.098     1   
243      18.780 3.000  Female     No  Thur  Dinner     2     0.160     1   

    bin_total  
0     (3, 18]  
1     (3, 18]  
2    (18, 35]  
3    (18, 35]  
4    (18, 35]  
..        ...  
239  (18, 35]  
240  (18, 35]  
241  (18, 35]  
242   (3, 18]  
243  (18, 35]  

[244 rows x 10 columns]

In [73]:
df.groupby(['time','bin_total'])[['ones']].count()

ones
time   bin_total      
Lunch  (3, 18]      45
       (18, 35]     21
       (35, 60]      2
Dinner (3, 18]      82
       (18, 35]     80
       (35, 60]     14

In [74]:
df.groupby(['time','bin_total'])[['ones']].count().groupby(level=0).apply(

lambda x:
    x/ x.sum()*100
)

ones
time   bin_total       
Lunch  (3, 18]   66.176
       (18, 35]  30.882
       (35, 60]   2.941
Dinner (3, 18]   46.591
       (18, 35]  45.455
       (35, 60]   7.955